### ISSUE TITLE
please detect parsing of ls output

### ISSUE THREAD

**Issue by pabs3** [koalaman/shellcheck#272]
Please detect parsing of ls output:

http://mywiki.wooledge.org/ParsingLs

**Comment by ormaaj** [comment]
It does. Always has afair.

    $ shellcheck -s sh - <<<'for x in $(ls); do :; done'

    In - line 1:
    for x in $(ls); do :; done
    ^-- SC2034: x appears unused. Verify it or export it.
             ^-- SC2045: Iterating over ls output is fragile. Use globs.

What shellcheck does not do is detect field splitting and globs resulting from unquoted expansions in general.

    $ shellcheck -s bash - <<<'y="foo b*ar"; for x in $y; do : "$x"; done'
    <no hints>

At minimum shellcheck should suggest an array when using `bash`/`ksh`. For sh it should perhaps recommend a function and `for x` iteration. Even better would be a heuristic to try and tailor the warning based upon the `set -f` state and `IFS` state but that would of course only be a guess.

Or, try and detect obviously correct usages:

    # This person clearly knows what they're doing, so don't suggest anything.
    for x in $(printf '%q ' "$@"); do ...

**Comment by ormaaj** [comment]
Actually it looks like shellcheck does do this in limited cases as discussed in #267. It just doesn't apply the same rule to every context where arguments are relevant.

```bash
# Works for simple commands.
 ~ $ shellcheck -s bash - <<<'y="foo b*ar"; cmd $y'

In - line 1:
y="foo b*ar"; cmd $y
                  ^-- SC2086: Double quote to prevent globbing and word splitting.

# Doesn't work when the first word isn't literal.
 ~ $ shellcheck -s bash - <<<'y="foo b*ar"; $y'

# Doesn't work for array assignments.
 ~ $ shellcheck -s bash - <<<'y="foo b*ar"; a=($y)'

In - line 1:
y="foo b*ar"; a=($y)
              ^-- SC2034: a appears unused. Verify it or export it.

# Doesn't work in loops.
 ~ $ shellcheck -s bash - <<<'y="foo b*ar"; for x in $y; do : "$x"; done'
```

**Comment by koalaman** [comment]
# This person clearly knows what they're doing, so don't suggest anything.
    for x in $(printf '%q ' "$@"); do ...

They demonstrably do not :P  The quoting added by printf %q will not be considered literal

Most of the places where shellcheck doesn't warn about splitting are by design, including the ones listed here (though the array assignment is still up for debate). 

If there are any cases involving either parsing ls or quoting that shellcheck should be handling, please reopen!

**Comment by ormaaj** [comment]
True I gave the wrong example. The  `%q` exception applies when storing an escaped argument list that will be used in an eval-like context that would otherwise trigger the hypothetical "unquoted expansion" warning.

**Comment by koalaman** [comment]
Fair enough, but even if you did `eval ... $var` you would still have to quote "$var"
